### ISSUE TITLE
Hi ! i have a question about Reprogramming Interpretation

### ISSUE THREAD

**Issue by Jeong-Eul** [KimMeen/Time-LLM#132]
How did you obtain the word set, and how did you create a visualization that shows how closely it is related to each prototype?

In the code, vocabulary is used by taking the weight of the LLM's embedding layer, but I thought it would be difficult to interpret which word combination prototype and which time series patch had the high attention score.

Thank you for reading my question.

**Comment by Zll99** [comment]
> How did you obtain the word set, and how did you create a visualization that shows how closely it is related to each prototype?
> 
> In the code, vocabulary is used by taking the weight of the LLM's embedding layer, but I thought it would be difficult to interpret which word combination prototype and which time series patch had the high attention score.
> 
> Thank you for reading my question.

Hi, there. I got the same confusion when I read their paper. Do you find some cues to solve this question?

**Comment by rlaals7349** [comment]
Hi, there. i have same confusion too. i can't find codes which create highly related text prototypes( (f) of figure 5 in this paper ). do you find any solutions ?

**Comment by fanyuqian0815** [comment]
> 嘿，你好。我也有同样的困惑。我找不到创建高度相关的文本原型的代码（本文图 5 的 （f））。您找到任何解决方案吗？

请问您找到了吗

**Comment by rlaals7349** [comment]
> > 嘿，你好。我也有同样的困惑。我找不到创建高度相关的文本原型的代码（本文图 5 的 （f））。您找到任何解决方案吗？
> 
> 请问您找到了吗

No Yet

**Comment by Jeong-Eul** [comment]
I think they used the weight matrix of the text prototype layer (self.mapping_layer = nn.Linear(self.vocab_size, self.num_tokens), where the shape of the weight matrix is (n_text_prototype, n_vocab)).

For interpretation, they may have used a predefined word set, which is a custom word set designed to exhibit various characteristics. For instance, it could include words that are highly related to time series or words that are less related to time series.

These words would then be encoded using a tokenizer to obtain their corresponding indices. Using these indices, they could extract the top N text prototypes by indexing the weight matrix.

The detailed process and results are as follows:

```python
wordset_1 = ['late', 'early', 'down', 'up', 'steady', 'short', 'long', 'peak', 'fluctuate', 'stabilize', 'irregular', 'consistent', 'sudden', 'gradual', 'missing', 'Not measured', 'nonlinear', 'linear', 'padding']

custom_word_indices = [tokenizer.encode(word, add_special_tokens=False)[0] for word in wordset_1]

custom_word_weights = weight_matrix[:, custom_word_indices]  # Shape: (1000, len(custom_word_set))

custom_word_weight_sums = np.sum(custom_word_weights, axis=1)
top_dimensions = custom_word_weight_sums.argsort()[-10:][::-1]

top_prototypes = custom_word_weights[top_dimensions]  # top 10 text prototype
print(top_dimensions)

plt.figure(figsize=(10, 6))
sns.heatmap(top_prototypes, cmap='RdPu', cbar=False)

plt.xlabel('Word Set 1')
plt.ylabel('Top Prototypes')
plt.savefig('wordset1.png', dpi=300)
plt.close()
```

![image](https://github.com/user-attachments/assets/4a2b8207-d84e-43cd-a1df-444dce9570a7)

**Comment by rlaals7349** [comment]
> I think they used the weight matrix of the text prototype layer (self.mapping_layer = nn.Linear(self.vocab_size, self.num_tokens), where the shape of the weight matrix is (n_text_prototype, n_vocab)).
> 
> For interpretation, they may have used a predefined word set, which is a custom word set designed to exhibit various characteristics. For instance, it could include words that are highly related to time series or words that are less related to time series.
> 
> These words would then be encoded using a tokenizer to obtain their corresponding indices. Using these indices, they could extract the top N text prototypes by indexing the weight matrix.
> 
> The detailed process and results are as follows:
> 
> ```python
> wordset_1 = ['late', 'early', 'down', 'up', 'steady', 'short', 'long', 'peak', 'fluctuate', 'stabilize', 'irregular', 'consistent', 'sudden', 'gradual', 'missing', 'Not measured', 'nonlinear', 'linear', 'padding']
> 
> custom_word_indices = [tokenizer.encode(word, add_special_tokens=False)[0] for word in wordset_1]
> 
> custom_word_weights = weight_matrix[:, custom_word_indices]  # Shape: (1000, len(custom_word_set))
> 
> custom_word_weight_sums = np.sum(custom_word_weights, axis=1)
> top_dimensions = custom_word_weight_sums.argsort()[-10:][::-1]
> 
> top_prototypes = custom_word_weights[top_dimensions]  # top 10 text prototype
> print(top_dimensions)
> 
> plt.figure(figsize=(10, 6))
> sns.heatmap(top_prototypes, cmap='RdPu', cbar=False)
> 
> plt.xlabel('Word Set 1')
> plt.ylabel('Top Prototypes')
> plt.savefig('wordset1.png', dpi=300)
> plt.close()
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/122766824/390969745-4a2b8207-d84e-43cd-a1df-444dce9570a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI4NTg5MjYsIm5iZiI6MTczMjg1ODYyNiwicGF0aCI6Ii8xMjI3NjY4MjQvMzkwOTY5NzQ1LTRhMmI4MjA3LWQ4NGUtNDNjZC1hMWRmLTQ0NGRjZTk1NzBhNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEyOVQwNTM3MDZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xN2JhNmQzMzgzNDY2Y2FiOGFmYjNiYzE0ZGI5ODljMzM5YThhYmIyYTU1MDZlZTdmNzI1Y2U0NGQzMDAzMmZmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.cEam1CycD1fQ-xxu328beV_52xWKCBPx50pKDzUbeSM)

for my opinion, i thought a little bit different way.(but your approach make sense more)

1. train a TimeLLM
2. choose top 10 words of similarity between (source embedding(from mapping layer) <-> word embedding)
   - for example cosine similarity
   - we can do it when using 4096 dimension from word embdding and 4096 dimension from transposed source embedding
   - if we want to draw (f) of fig 5, we have to num_tokens = 10, because y axis is 0~9 ( or truncate a few prototypes )
   - we can get a 10 different pair of similarities. ( each 1 prototypes with top 10 words )
3. draw heatmap about all of different pairs (proto and 10 words)


and a result like below (draw only single learned text prototype and top 10 words )

![image](https://github.com/user-attachments/assets/2078771e-ad33-4d2c-a6c0-e33d9d8e2654)

+ we tried to extract original words using tokenizer from word embeddings which calculated top 10 similar to prototypes 
and the result is below

![image](https://github.com/user-attachments/assets/dc126843-b6eb-4a6f-9a67-74fe73a9a98c)

i'm confusing it is correct way to create (f) of figure 5

**Comment by Jeong-Eul** [comment]
rlaals7349님 한국인이신 것 같네요. 한국말로 할게요. 저도 처음에는 사전에 단어 집합을 구성 해놓지 않고, attention score가 높았던 상위 10개 prototype에 대해서 tokenizer index 기준으로 가장 연관 있는 단어들을 뽑는 식으로 진행했었는데, rlaals7349님 처럼 정말 말도 안되는 단어들이 나와서 당황했습니다. 제 생각에는 reprogramming이라는게 아무래도 노이즈를 학습하는 과정이다 보니 관련 없는 단어들도 참여하게 되고, 결과적으로 시계열 패치 하나가 일반적으로 시계열과 관련있는 단어들만 섞이는게 아닌 것으로 이해하고 넘어갔었습니다. 논문에서는 word set들이 시계열과 관련있는 단어와 그렇지 않은 것들을 보줬는데, 사실 그렇게 와닿지는 않는 결과입니다. 그래도 논문에서 한 방식처럼 하게되면 시계열과 관련있는 단어들과 그렇지 않은 단어들 사이의 히트맵이 다르긴해서 어느 정도 말은 된다고 생각하고 넘어갔습니다. 고생하세요 :)

<i>[English]</i>

At first, I also tried an approach where I didn’t predefine a word set but instead selected the most related words based on the tokenizer index for the top 10 prototypes with the highest attention scores. However, I was taken aback when completely nonsensical words emerged, just like what happened to you, rlaals7349. I think this might be because reprogramming inherently involves learning noise, which leads to unrelated words being included as well. Consequently, I interpreted it as each time-series patch not necessarily being mixed with words solely related to the time series in a typical sense. In the paper, the word sets included a mix of words that were both related and unrelated to the time series, but honestly, the results didn’t resonate strongly with me. Still, when following the method described in the paper, the heatmaps between words related to the time series and unrelated words do show some differences, so I decided it made sense to some extent and moved on.

**Comment by rlaals7349** [comment]
> rlaals7349님 한국인이신 것 같네요. 한국말로 할게요. 저도 처음에는 사전에 단어 집합을 구성 해놓지 않고, attention score가 높았던 상위 10개 prototype에 대해서 tokenizer index 기준으로 가장 연관 있는 단어들을 뽑는 식으로 진행했었는데, rlaals7349님 처럼 정말 말도 안되는 단어들이 나와서 당황했습니다. 제 생각에는 reprogramming이라는게 아무래도 노이즈를 학습하는 과정이다 보니 관련 없는 단어들도 참여하게 되고, 결과적으로 시계열 패치 하나가 일반적으로 시계열과 관련있는 단어들만 섞이는게 아닌 것으로 이해하고 넘어갔었습니다. 논문에서는 word set들이 시계열과 관련있는 단어와 그렇지 않은 것들을 보줬는데, 사실 그렇게 와닿지는 않는 결과입니다. 그래도 논문에서 한 방식처럼 하게되면 시계열과 관련있는 단어들과 그렇지 않은 단어들 사이의 히트맵이 다르긴해서 어느 정도 말은 된다고 생각하고 넘어갔습니다. 고생하세요 :)
> 
> [English]
> 
> At first, I also tried an approach where I didn’t predefine a word set but instead selected the most related words based on the tokenizer index for the top 10 prototypes with the highest attention scores. However, I was taken aback when completely nonsensical words emerged, just like what happened to you, rlaals7349. I think this might be because reprogramming inherently involves learning noise, which leads to unrelated words being included as well. Consequently, I interpreted it as each time-series patch not necessarily being mixed with words solely related to the time series in a typical sense. In the paper, the word sets included a mix of words that were both related and unrelated to the time series, but honestly, the results didn’t resonate strongly with me. Still, when following the method described in the paper, the heatmaps between words related to the time series and unrelated words do show some differences, so I decided it made sense to some extent and moved on.

빠른 답변 감사드립니다. 아무래도 말씀하신 것처럼, 그렇게 접근하는 것이 맞는 것 같아요. 아래처럼 다른 LLM 논문을 살펴보더라도 wordset을 구성하고, time series와 관련이 높은 단어들이 높은 score를 가지고, 그렇지 않은 단어들이 낮은 score를 가진다고 해석을 하더라고요. 덕분에 많이 도움이 되었습니다. 감사합니다.

[English]
Thank you for the quick response. As you mentioned, it seems that approach is indeed the right one. Looking at other LLM papers, they also interpret it as constructing a word set where words closely related to time series have higher scores, while unrelated words have lower scores. Your comment has been very helpful. Thank you!

![image](https://github.com/user-attachments/assets/5a122de6-2fbc-45cc-be64-9a0d0bf5dac7)

**Comment by fanyuqian0815** [comment]
> 我认为他们使用了文本原型层的权重矩阵 （self.mapping_layer = nn.Linear（self.vocab_size， self.num_tokens），其中权重矩阵的形状为 （n_text_prototype， n_vocab））。
> 
> 为了进行解释，他们可能使用了预定义的词集，这是一个旨在展示各种特征的自定义词集。例如，它可以包括与时间序列高度相关的单词或与时间序列不太相关的单词。
> 
> 然后使用分词器对这些单词进行编码，以获得它们相应的索引。使用这些索引，他们可以通过索引权重矩阵来提取前 N 个文本原型。
> 
> 详细过程和结果如下：
> 
> ```python
> wordset_1 = ['late', 'early', 'down', 'up', 'steady', 'short', 'long', 'peak', 'fluctuate', 'stabilize', 'irregular', 'consistent', 'sudden', 'gradual', 'missing', 'Not measured', 'nonlinear', 'linear', 'padding']
> 
> custom_word_indices = [tokenizer.encode(word, add_special_tokens=False)[0] for word in wordset_1]
> 
> custom_word_weights = weight_matrix[:, custom_word_indices]  # Shape: (1000, len(custom_word_set))
> 
> custom_word_weight_sums = np.sum(custom_word_weights, axis=1)
> top_dimensions = custom_word_weight_sums.argsort()[-10:][::-1]
> 
> top_prototypes = custom_word_weights[top_dimensions]  # top 10 text prototype
> print(top_dimensions)
> 
> plt.figure(figsize=(10, 6))
> sns.heatmap(top_prototypes, cmap='RdPu', cbar=False)
> 
> plt.xlabel('Word Set 1')
> plt.ylabel('Top Prototypes')
> plt.savefig('wordset1.png', dpi=300)
> plt.close()
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/122766824/390969745-4a2b8207-d84e-43cd-a1df-444dce9570a7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQyNzg2MzksIm5iZiI6MTczNDI3ODMzOSwicGF0aCI6Ii8xMjI3NjY4MjQvMzkwOTY5NzQ1LTRhMmI4MjA3LWQ4NGUtNDNjZC1hMWRmLTQ0NGRjZTk1NzBhNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjE1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIxNVQxNTU4NTlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0zODgwNjc1NmU2NmRlYjlkMjZkNDBiODdlNGRkMDJhMmQzNjEwNThjZmE0NWYzNjYxMzcwYmVlNmVmNmU2YTk3JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.eakVik4boMS9m6fkwn0pKjzLNBhrrHD2TS5p-pRVT3c)

请问预定义的词集我们可以输出看到是什么样子的吗

**Comment by fanyuqian0815** [comment]
> rlaals7349님 한국인이신 것 같네요.한국말로 할게요.저도 처음에는 사전에 단어 집합을 구성 해놓지 않고， attention score가 높았던 상위 10개 prototype에 대해서 tokenizer index 기준으로 가장 연관 있는 단어들을 뽑는 식으로 진행했었는데， rlaals7349님 처럼 정말 말도 안되는 단어들이 나와서 당황했습니다.제 생각에는 reprogramming이라는게 아무래도 노이즈를 학습하는 과정이다 보니 관련 없는 단어들도 참여하게 되고， 결과적으붼되 �� 시계열 패치 하나가 일반적으로 시계열과 관련있는 단어들만 섞이는게 아닌 것으로 이해하고 넘어갔었습니다.논문에서는 word set들이 시계열과 관련있는 단어와 그렇지 않은 것들을 보줬는데， 사실 그렇게 와닿지는 않는 결과입니다.그래도 논문에서 한 방식처럼 하게되면 시계열과 관련있는 단어들과 그렇지 않은 단어들 사이의 히트맵이 다르긴해서 어느 정도 말은 된다고 생각하고 넘어갔습니다.고생하세요 :)
> 
> [中文]
> 
> 起初，我还尝试了一种方法，我没有预定义单词集，而是根据分词器索引为注意力得分最高的前 10 个原型选择最相关的单词。然而，当完全荒谬的词出现时，我吓了一跳，就像发生在你身上的事情一样，rlaals7349。我认为这可能是因为重新编程本身就涉及学习噪声，这导致不相关的单词也被包括在内。因此，我将其解释为每个时间序列补丁，而不一定与典型意义上的时间序列仅相关的单词混合在一起。在论文中，词集包括与时间序列相关和无关的单词混合，但老实说，结果并没有引起我的强烈共鸣。尽管如此，当遵循论文中描述的方法时，与时间序列相关的单词和不相关的单词之间的热图确实显示出一些差异，因此我认为这在某种程度上是有意义的，然后继续前进。

请问预定义的词集我们可以输出看到是什么样子的吗
